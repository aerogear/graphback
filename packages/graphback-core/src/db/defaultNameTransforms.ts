export type DatabaseNameTransformDirection = 'from-db' | 'to-db'

/**
 * Transform to/from database table or column name
 */
export type DatabaseNameTransform = (
  name: string,
  direction: DatabaseNameTransformDirection,
) => string

/**
 * Transform to/from database table name
 * 
 * @param name - model name
 * @param direction - transform to or from database
 */
export function defaultTableNameTransform(name: string, direction: DatabaseNameTransformDirection) {
  if (direction === 'to-db') {
    return name.toLowerCase();
  }

  return name
}

export function transformForeignKeyName(name: string, direction: DatabaseNameTransformDirection = 'to-db') {
  if (direction === 'to-db') {
    return `${name}Id`;
  }

  return name;
}