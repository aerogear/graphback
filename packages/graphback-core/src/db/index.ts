export * from './defaultNameTransforms';
export * from './modelTableMapper';
export * from './dataMapper';
export * from './getPrimaryKey';
