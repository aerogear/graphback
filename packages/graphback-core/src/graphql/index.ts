
export * from "./GraphQLContextCreator"
export * from "./graphqlUtils"
export * from "./directives"