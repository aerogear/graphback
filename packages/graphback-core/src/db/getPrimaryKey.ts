import { GraphQLField, GraphQLObjectType, getNamedType, isScalarType, GraphQLInputField } from "graphql";
import { parseMetadata } from 'graphql-metadata';

/**
 * Returns the primary key field of a GraphQL object.
 * First looks for the existence of a `@id` field annotation,
 * otherwise tries to find an `id: ID` field.
 *
 * @param graphqlType
 */
export function getPrimaryKey(graphqlType: GraphQLObjectType): GraphQLField<any, any> {
  const fields = Object.values(graphqlType.getFields());

  const autoPrimaryKeyFromScalar: GraphQLField<any, any>[] = [];
  let primaryKey: GraphQLField<any, any>;
  let primariesCount = 0;
  for (const field of fields) {
    const hasIdMarker = parseMetadata("id", field);
    if (hasIdMarker) {
      primaryKey = field;
      primariesCount += 1;
    } else if (isAutoPrimaryKey(field)) {
      autoPrimaryKeyFromScalar.push(field);
    }
  }

  if (primariesCount > 1) {
    throw new Error(`${graphqlType.name} type should not have multiple '@id' annotations.`)
  }

  if (primaryKey) {
    return primaryKey;
  }

  if (autoPrimaryKeyFromScalar.length > 1) {
    throw new Error(`${graphqlType.name} type should not have two potential primary keys: "_id" and "id". Use '@id' annotations to indicate which one is to be used.`);
  }

  primaryKey = autoPrimaryKeyFromScalar.shift();

  if (!primaryKey) {
    throw new Error(`${graphqlType.name} type has no primary field.`)
  }

  return primaryKey;
}

/**
 * Check if a GraphQLField can be inferred as a primary key, specific for each database:
 * A field is a potential primary key if:
 * - is named "id" and has type "ID", auto increment primary key for relational database
 * - is named "_id" and has scalar type "GraphbackObectID", a BSON primary key for MongoDB
 * @param field
 */
export function isAutoPrimaryKey(field: GraphQLField<any, any> | GraphQLInputField): boolean {
  const { type, name: fieldName } = field;
  const baseType = getNamedType(type);
  const name = baseType.name;

  return ((fieldName === 'id' && name === 'ID') || (fieldName === "_id" && name === "GraphbackObjectID")) && isScalarType(baseType);
}
