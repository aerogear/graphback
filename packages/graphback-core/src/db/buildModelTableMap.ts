import { GraphQLField, GraphQLFieldMap, GraphQLObjectType } from "graphql";
import { parseDbAnnotations } from '../annotations/parser';
import { defaultTableNameTransform } from './defaultNameTransforms';
import { getPrimaryKey } from './getPrimaryKey';

/**
 * Contains mapping information between GraphQL Model type and database table
 * 
 * - typeName: Original GraphQLObjectType name
 * - tableName: Name of datase table
 * - id: Indicates the primary key field
 * - fieldMap: Object of key-value mapping between GraphQL fields and database columns.
 */
export interface ModelTableMap {
  typeName: string
  tableName: string
  idField: string
  fieldMap?: { [key: string]: string }
}

/**
 * Gets the datase column name for a GraphQL type.
 * Checks for the `@db.name` annotation for a customised name
 * 
 * @param field 
 */
export function getTableName(model: GraphQLObjectType): string {
  let tableName = defaultTableNameTransform(model.name, 'to-db');

  const dbAnnotations = parseDbAnnotations(model);
  if (dbAnnotations.name) {
    tableName = dbAnnotations.name;
  }

  return tableName;
}

/**
 * Gets the datase column name for a GraphQL field.
 * Checks for the `@db.name` annotation for a customised name
 * 
 * @param field 
 */
export function getColumnName(field: GraphQLField<any, any>): string {
  let columnName = field.name;

  const dbAnnotations = parseDbAnnotations(field);
  if (dbAnnotations.name) {
    columnName = dbAnnotations.name;
  }

  return columnName;
}

function mapFieldsToColumns(fieldMap: GraphQLFieldMap<any, any>): any {
  return Object.values(fieldMap).reduce((obj: any, field: GraphQLField<any, any>) => {
    const columnName = getColumnName(field);

    if (field.name !== columnName) {
      obj[field.name] = columnName;
    }

    //TODO: Map relationship fields

    return obj;
  }, {});
}

/**
 * Builds a database mapping model of a GraphQLObject type.
 * @param model - The GraphQL object data model representation
 * 
 * @returns {ModelTableMap} A model containing the table name, any field customisations and a mapping of the primary key field.
 */
export const buildModelTableMap = (model: GraphQLObjectType): ModelTableMap => {
  const primaryKeyField = getPrimaryKey(model);
  const tableName = getTableName(model);
  const fieldMap = mapFieldsToColumns(model.getFields());

  return {
    idField: getColumnName(primaryKeyField),
    typeName: model.name,
    tableName,
    fieldMap
  }
}