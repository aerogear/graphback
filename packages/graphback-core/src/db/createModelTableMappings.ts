import { GraphQLField, GraphQLFieldMap, GraphQLObjectType, GraphQLSchema } from "graphql";
import { parseDbAnnotations } from '../annotations/parser';
import { getUserModels } from '../crud';
import { getModelTypesFromSchema } from '../plugin/getModelTypesFromSchema';
import { DatabaseNameTransformDirection, defaultColumnNameTransform, defaultTableNameTransform } from './defaultNameTransforms';
import { getPrimaryKey } from './getPrimaryKey';

/**
 * Contains mapping information between GraphQL Model type and database table
 * 
 * - typeName: Original GraphQLObjectType name
 * - tableName: Name of datase table
 * - id: Indicates the primary key field
 * - fieldMap: Object of key-value mapping between GraphQL fields and database columns.
 */
export interface ModelTableMapping {
  typeName: string
  tableName: string
  idField: string
  fieldMap?: any
}

export const getModelTableMappings = (schema: GraphQLSchema): ModelTableMapping[] => {
  const modelTypes = getModelTypesFromSchema(schema);
  const models = getUserModels(modelTypes);

  return mapModelsToTables(models);
}

export function getModelTableMapping(modelName: string, mappings: ModelTableMapping[]) {
  return mappings.find((m: ModelTableMapping) => m.typeName === modelName);
}

function mapModelsToTables(models: GraphQLObjectType[]): ModelTableMapping[] {
  return models.map((model: GraphQLObjectType) => {
    return {
      idField: getPrimaryKey(model),
      typeName: model.name,
      tableName: getTableOrColumnName(model),
      fieldMap: mapFieldsToColumns(model.getFields())
    }
  });
}

function mapFieldsToColumns(fieldMap: GraphQLFieldMap<any, any>) {
  let hasCustomMappings = false;
  const mappedFields = Object.values(fieldMap).reduce((obj: any, field: GraphQLField<any, any>) => {

    const columnName = getTableOrColumnName(field);

    if (field.name !== columnName) {
      hasCustomMappings = true;
    }

    obj[field.name] = columnName;

    return obj;
  }, {});

  if (!hasCustomMappings || !Object.keys(mappedFields).length) {
    return undefined;
  }

  return mappedFields;
}

export function getTableOrColumnName(modelOrField: GraphQLObjectType | GraphQLField<any, any>): string {
  let transformedName = transformTableOrColumnName(modelOrField);

  const dbAnnotations = parseDbAnnotations(modelOrField);

  if (dbAnnotations.name) {
    transformedName = dbAnnotations.name;
  }

  return transformedName;
}

function transformTableOrColumnName(modelOrField: GraphQLObjectType | GraphQLField<any, any>, direction: DatabaseNameTransformDirection = 'to-db'): string {
  if (modelOrField instanceof GraphQLObjectType) {
    return defaultTableNameTransform(modelOrField.name, direction);
  }

  return defaultColumnNameTransform(modelOrField.name, direction);
}