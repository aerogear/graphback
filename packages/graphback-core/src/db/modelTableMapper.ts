import { GraphQLField, GraphQLFieldMap, GraphQLObjectType, GraphQLSchema } from "graphql";
import { parseDbAnnotations } from '../annotations/parser';
import { getUserModels } from '../crud';
import { getModelTypesFromSchema } from '../plugin/getModelTypesFromSchema';
import { DatabaseNameTransformDirection, defaultColumnNameTransform, defaultTableNameTransform } from './defaultNameTransforms';
import { getPrimaryKey } from './getPrimaryKey';

/**
 * Contains mapping information between GraphQL Model type and database table
 * 
 * - typeName: Original GraphQLObjectType name
 * - tableName: Name of datase table
 * - id: Indicates the primary key field
 * - fieldMap: Object of key-value mapping between GraphQL fields and database columns.
 */
export interface ModelTableMap {
  typeName: string
  tableName: string
  idField: string
  fieldMap?: any
}

export const findModelTableMappings = (schema: GraphQLSchema): ModelTableMap[] => {
  const modelTypes = getModelTypesFromSchema(schema);
  const models = getUserModels(modelTypes);

  return mapModelsToTables(models);
}

export function getModelTableMapping(mappedType: GraphQLObjectType) {
  // TODO 
  // return mappings.find((m: ModelTableMapping) => m.typeName === modelName);
}

function mapModelsToTables(models: GraphQLObjectType[]): ModelTableMap[] {
  return models.map((model: GraphQLObjectType) => {
    return {
      idField: getPrimaryKey(model).name,
      typeName: model.name,
      tableName: getTableOrColumnName(model),
      fieldMap: mapFieldsToColumns(model.getFields())
    }
  });
}

function mapFieldsToColumns(fieldMap: GraphQLFieldMap<any, any>): any {
  return Object.values(fieldMap).reduce((obj: any, field: GraphQLField<any, any>) => {
    const columnName = getTableOrColumnName(field);

    if (field.name !== columnName) {
      obj[field.name] = columnName;
    }


    return obj;
  }, {});
}

export function getTableOrColumnName(modelOrField: GraphQLObjectType | GraphQLField<any, any>): string {
  let transformedName = getMappedDatabaseName(modelOrField);

  const dbAnnotations = parseDbAnnotations(modelOrField);

  if (dbAnnotations.name) {
    transformedName = dbAnnotations.name;
  }

  return transformedName;
}

function getMappedDatabaseName(modelOrField: GraphQLObjectType | GraphQLField<any, any>, direction: DatabaseNameTransformDirection = 'to-db'): string {
  if (modelOrField instanceof GraphQLObjectType) {
    return defaultTableNameTransform(modelOrField.name, direction);
  }

  return defaultColumnNameTransform(modelOrField.name, direction);
}

export function findModelTableMap(name: string, mappings: ModelTableMap[]) {
  return mappings.find((m: ModelTableMap) => m.typeName === name);
}