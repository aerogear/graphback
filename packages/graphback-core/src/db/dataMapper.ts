import { ModelTableMap } from './buildModelTableMap';

export interface TableDataMap {
  idField?: TableID
  table?: string
  data?: any
  fieldMap?: any
}

export interface TableID {
  name: string
  value?: any
}

function getTableId(idField: string, data: any): TableID {
  if (!idField) { return undefined };

  let value: any;
  if (data && data[idField]) {
    value = data[idField];
  }

  return {
    name: idField,
    value
  }
}

export const getDatabaseArguments = (modelMap: ModelTableMap, data?: any): TableDataMap => {
  const idField = modelMap.idField;
  // Transform data if it defined
  let transFormedData: any;
  if (data) {
    const keys = Object.keys(data);
    transFormedData = {};
    for (const key of keys) {
      const value: any = data[key];
      transFormedData[key] = value && typeof value === 'object' ? JSON.stringify(value) : value;
    }
  };

  return {
    idField: getTableId(idField, data),
    data: transFormedData
  }
}
