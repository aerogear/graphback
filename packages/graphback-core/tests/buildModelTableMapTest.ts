//tslint:disable-next-line: match-default-export-name no-implicit-dependencies
import { buildSchema, GraphQLObjectType } from 'graphql';
import { buildModelTableMap } from '../src/db/buildModelTableMap';

test('should build model-table mapping using default values', () => {
    const schema = buildSchema(`
    """
    @model
    """
    type User {
        id: ID!
        email: String!
    }`);

    const userModel = schema.getType("User") as GraphQLObjectType;

    const modelTableMap = buildModelTableMap(userModel);

    expect(modelTableMap.tableName).toEqual('user');
    expect(modelTableMap.fieldMap).toEqual({});
});

test('should build mapping using custom values from mapping annotations', () => {
    const schema = buildSchema(`
    """
    @model
    @db.name: 'user_account'
    """
    type User {
        id: ID!
        """
        @db.name: 'user_email'
        """
        email: String!
        """
        @db.name: 'userName'
        """
        name: String!
        address: String
        """
        @db.name: 'AccountConfirmed'
        """
        accountConfirmed: Boolean
    }`);

    const userModel = schema.getType("User") as GraphQLObjectType;

    const modelTableMap = buildModelTableMap(userModel);

    expect(modelTableMap.tableName).toEqual('user_account');
    expect(modelTableMap.fieldMap).toEqual(
        { email: 'user_email', name: 'userName', accountConfirmed: 'AccountConfirmed' }
    );
    expect(JSON.stringify(modelTableMap, undefined, 1)).toMatchSnapshot()
});

test('should use default ID field', () => {
    const schema = buildSchema(`
    """
    @model
    """
    type User {
        id: ID!
        secondaryID: ID!
        email: String!
    }`);

    const userModel = schema.getType("User") as GraphQLObjectType;

    const modelTableMap = buildModelTableMap(userModel);

    expect(modelTableMap.idField).toEqual('id');
});

test('should use custom ID field from annotation', () => {
    const schema = buildSchema(`
    """
    @model
    """
    type User {
        id: ID!
        """
        @id
        @db.name: 'user_email'
        """
        email: String!
    }`);

    const userModel = schema.getType("User") as GraphQLObjectType;

    const modelTableMap = buildModelTableMap(userModel);

    expect(modelTableMap.idField).toEqual('user_email');
});

test('should throw error if no ID Field', () => {
    const schema = buildSchema(`
    """
    @model
    """
    type User {
        primary: ID!
        name: String!
    }`);

    const userModel = schema.getType("User") as GraphQLObjectType;

    expect(() => buildModelTableMap(userModel)).toThrowError('User type has no primary field.');
});