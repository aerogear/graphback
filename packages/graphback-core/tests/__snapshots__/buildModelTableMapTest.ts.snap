// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build mapping using custom values from mapping annotations 1`] = `
"{
 \\"idField\\": \\"id\\",
 \\"typeName\\": \\"User\\",
 \\"tableName\\": \\"user_account\\",
 \\"fieldMap\\": {
  \\"email\\": \\"user_email\\",
  \\"name\\": \\"userName\\",
  \\"accountConfirmed\\": \\"AccountConfirmed\\"
 }
}"
`;
